Added preliminary HuC-3 documentation.

cuavas opened the pull request:
This adds preliminary documentation for the HuC-3 combination memory controller, real-time clock, tone generator and infrared communication controller.  Most of the information is already collected in the [cited Gameboy Development Forum thread](https://gbdev.gg8.se/forums/viewtopic.php?id=744) as well as various emulator sources.

I also edited the HuC1 page a little after looking at the contributing guidelines:
* Use `$` as prefix for hexadecimal numbers
* Use no prefix for hexadecimal addresses
* Use en dash rather than hyphen to denote ranges
* Hyphenate “read-only” and “write-only”

The MBC5 page change just wraps a line that stood out as excessively long at 72 columns like the rest of the file seems to use.

ISSOtm commented on src/HuC1.md:
```diff
@@ -1,50 +1,51 @@
-on, 0x00 turns it off. Read from this region to see either 0xC1 (saw
-light) or 0xC0 (did not see light). When in \"RAM mode\" (wrote
-something other than 0x0E to 0x000) this region behaves like normal cart
+When in \"IR mode\" (wrote $0E to 0000), the IR register is visible
```
```suggestion
When in \"IR mode\" (wrote $0E to $0000), the IR register is visible
```

cuavas replied:
Well, I took the following from CONTRIBUTING.md to heart:

> To prevent clutter, don't use a prefix for hex numbers when it's clear from the context that a number is hexadecimal. For example, addresses and lists of opcodes. In those cases, zero-pad, even for smaller numbers: `0000-3FFF` instead of `0-3FFF`.

I padded addresses to four hexadecimal digits and didn’t add a prefix, as addresses are implied to be hexadecimal.  Should this rule be updated/clarified?

ISSOtm replied:
Fair that you're following the rules; I think that addresses in running text like this should get a hex prefix at least for consistency ("0E" is arguably unambiguous), and the prefix helps it stand out from other words.

ISSOtm commented on src/HuC1.md:
```diff
@@ -1,50 +1,51 @@
+here. Write to this region to control the IR transmitter. $01 turns it
+on, $00 turns it off. Read from this region to see either $C1 (saw
+light) or $C0 (did not see light). When in \"RAM mode\" (wrote
+something other than $0E to 0000) this region behaves like normal cart
```
```suggestion
something other than $0E to $0000) this region behaves like normal cart
```

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+Games write $01 here on startup, but it doesn\'t seem to have any
+observable effect.
+
+
```
```suggestion
```

cuavas replied:
I’ve intentionally left two blank lines before every level 2 heading (`##` prefix) besides the first to make the sections visible when editing the source in a text editor.

ISSOtm replied:
AFAIK, that is not valid Markdown (and I have seen parsers act incorrectly in that case), so I'd rather avoid it.

cuavas replied:
You know, I really hate markdown because every single implementation of it is different – it’s the opposite of a standard.  I’ll remove them here.

ISSOtm replied:
Well, that's what [CommonMark](https://commonmark.org) attempts to fix, but bugs are legion.

cuavas replied:
> Well, that's what [CommonMark](https://commonmark.org/) attempts to fix, but bugs are legion.

I’m not even sure it’s “bugs” as such – the different dialects are evolving in different directions for whatever reason.

For example GitHub markdown doesn’t require a blank line to introduce a list, treats unadorned newlines as hard line breaks, and so in.  It seems to be an intentional break with the other flavours.

RST is also evolving in its own direction, with its own syntax for links, anchors, invoking macros, etc.  It seems to be more influenced by the TeX/LaTeX family of document markup languages.  We end up needing to lean pretty heavily on some of its features for the MAME documentation.

Anyway, that’s all kind of beside the point.  All the differences just make me nervous about writing anything in markdown when it’s going to be rendered by an unfamiliar engine.

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+  58–5A  | Event time minutes
+  5B–5D  | Event time days
+
+
```
```suggestion
```

ISSOtm commented on src/MBC5.md:
```diff
@@ -29,8 +29,10 @@ Same as for MBC1, except that RAM sizes are 8 KiB, 32 KiB and 128 KiB.
 
-Actual MBCs actually enable RAM when writing any value whose bottom 4 bits equal $A (so $0A, $1A, and so on), and disable it when writing anything else.
-Relying on this behavior is not recommended for compatibility reasons.
+Actual MBCs actually enable RAM when writing any value whose bottom 4
+bits equal $A (so $0A, $1A, and so on), and disable it when writing
+anything else. Relying on this behavior is not recommended for
+compatibility reasons.
```
```suggestion
Actual MBCs actually enable RAM when writing any value whose bottom 4 bits equal $A (so $0A, $1A, and so on), and disable it when writing anything else.
Relying on this behavior is not recommended for compatibility reasons.
```
Text should not be wrapped at 80 columns (this is a hold-over that we haven't done a pass on); one line per sentence is preferred, as it keeps diffs clearer: a change does not reflow all text following it, muddying `git blame`.

cuavas replied:
> Text should not be wrapped at 80 columns (this is a hold-over that we haven't done a pass on); one line per sentence is preferred, as it keeps diffs clearer: a change does not reflow all text following it, muddying git blame.

Interesting.  We’ve made the opposite decision for MAME’s documentation.  We found that diffs were borderline impossible to deal with when paragraphs weren’t hard wrapped in the source.  I’ll back out this part of the change.

ISSOtm replied:
If paragraphs aren't wrapped at all, yes, it's a nightmare. This is why I think that "one line per sentence" (which I somewhat borrowed from man pages, where periods at EOL are treated as sentence end markers) is a good compromise.

ISSOtm commented on src/HuC1.md:
```diff
@@ -1,50 +1,51 @@
+  $A000–BFFF  | Cart RAM or IR register
 
-### 0000-1FFF IR Select (Write Only)
+### 0000–1FFF IR Select (Write-Only)
```
I appreciate the effort, but the section title does not follow [the style guide](https://github.com/gbdev/pandocs/blob/master/CONTRIBUTING.md#8-section-titles) anyway, so I am currently preparing a PR to fix that. (If you can follow it in the new docs, that would be very appreciated!)

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+real-time clock and tone generator are implemented using a 4-bit
+microcontroller core with internal program ROM.
+
+
```
```suggestion
```

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+below), this memory space is used to access an 8 KiB external RAM Bank,
+or a single I/O Register.
+
+
```
```suggestion
```

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+
+## I/O Registers
+
+These registers are accessed by reading or writing int the A000–BFFF
```
```suggestion
These registers are accessed by reading or writing in the $A000–BFFF
```

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+
+Similar to the equivalent register of the HuC-1. The least significant
+bit is used for infrared communication.
+
```
```suggestion
```

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+## RTC Communication Protocol
+
+The HuC-3 chip provides read and write access to a 256-nybble window
+into the RTC MCU\'s internal memory. Multi-nybble values are stored
```
Is the backslash necessary?

cuavas replied:
I saw it in use in other markdown documents in your repository (e.g. HuC1.md).  Presumably it’s to prevent automatic conversion of foot/inch marks to typographical quotes.  Should I just use typographical quote characters?

ISSOtm replied:
Yeah, those should be more appropriate here. I think we have some backslashes left over from the MediaWiki conversion, but in general, I'd rather not have them if they aren't necessary.

cuavas replied:
OK, I’ll fix it.

ISSOtm commented on src/HuC3.md:
```diff
@@ -0,0 +1,171 @@
+
+Games use the following sequence to execute a command:
+
+* Write $0D to 0000 (select RTC Semaphore register)
+* Poll A000 until least significant bit is set (wait until ready)
+* Write $0B to 0000 (select RTC Command/Argument register)
+* Write command and argument to A000
+* Write $0D to 0000 (select RTC semaphore register)
+* Write $FE to A000 (clear semaphore, requesting MCU execute command)
+* Poll A000 until least significant bit is set (wait for completion)
+* Write $0C to 0000 (select RTC Command/Response register)
+* Read A000 and use value from 4 least significant bits
```
```suggestion
* Write $0D to $0000 (select RTC Semaphore register)
* Poll $A000 until least significant bit is set (wait until ready)
* Write $0B to $0000 (select RTC Command/Argument register)
* Write command and argument to $A000
* Write $0D to $0000 (select RTC semaphore register)
* Write $FE to $A000 (clear semaphore, requesting MCU execute command)
* Poll $A000 until least significant bit is set (wait for completion)
* Write $0C to $0000 (select RTC Command/Response register)
* Read $A000 and use value from 4 least significant bits
```

cuavas replied:
Once again, I was following the rule I quoted from CONTRIBUTING.md that says to omit the radix prefix for addresses.

ISSOtm reviewed (changes requested):
Thanks for the doc! Considering this is called "preliminary", are you planning to expand this doc further?

cuavas commented:
> Considering this is called "preliminary", are you planning to expand this doc further?

Depends on how motivated I am to try and reverse-engineer the behaviour of a microcontroller embedded in a chip with very little to go on.  The fact is, knowledge about the HuC-3 is very limited at the moment.

ISSOtm commented:
That's perfectly fine, I was asking only to know whether to mark this PR as a draft :)

cuavas commented:
OK:
* Changed headings to use em dash separator between address and description, square brackets around read/write usage information, and lowercase for read/write usage information.
* Removed double blank lines.
* Fixed the stray “t”.
* Added radix prefix to addresses in prose.
* Backed out MBC5.md change.
* Replaced backslash-escaped foot/inch marks with typographical quotes/apostrophes.

I still don’t know exactly what this is actually going to look like in the end, as every flavour of markdown is handled differently.  I’m mostly familiar with Sphinx RST (as used in MAME’s documentation), and the GitHub and Reddit flavours.

cuavas commented:
> Are you unable to set up mdBook locally? I can upload a render myself if that's the case.

It’s a matter of being thin on the ground.  I have a constant backlog of MAME work, as well as real life™, and I’m severely sleep deprived.  I simply don’t have the energy to set up mdBook locally.

The GitHub workflows seemed to want to upload to your main site, so I didn’t try enabling them on my fork, because I don’t want to hit you with security alerts when it fails to push out the updated.

ISSOtm commented:
The GitHub workflow has a safeguard so that it only updates the website when ran as a push, to `master`, on `gbdev/pandocs`. Don't worry, we already had a few issues with *my* fork way back when. :P

Anyway, [I manually rendered the current state here](http://eldred.fr/pandocs/HuC3.html).

ISSOtm commented:
> I still don’t know exactly what this is actually going to look like in the end, as every flavour of markdown is handled differently. I’m mostly familiar with Sphinx RST (as used in MAME’s documentation), and the GitHub and Reddit flavours.

Note that [GH-flavored MarkDown](https://github.github.com/gfm/) is a couple of extensions over CommonMark, which [mdBook's parser](https://lib.rs/pulldown-cmark) implements, plus a few GitHub extensions.

I'll review tomorrow, I must ensure I don't become sleep-deprived myself ^^'

avivace commented on src/HuC1.md:
```diff
@@ -1,50 +1,51 @@
+HuC1 is an MBC used by some Game Boy games which, besides doing the
+usual MBC stuff, also provides infrared communication. A lot of sources
+on the internet said that HuC1 was “similar to MBC1”, but they didn’t
+provide any detail. I took a look, and it turns out that HuC1 differs
```
```suggestion
provide any detail. The HuC1 differs
```

Since we're here already, those kind of comments/remarks should be removed

cuavas replied:
This page is basically copied from a jrra.zone blog post (hopefully with permission): http://jrra.zone/blog/huc1.html

I wasn’t editing it for style, I don’t know whether the license of the blog post would even allow that.  There’s no clear copyright notice/license on the original blog post, so Berne Convention automatic copyright/all rights reserved applies.

ISSOtm replied:
All content under Pan Docs has been licensed to the public domain, so that should be fine.

cuavas replied:
I tightened up the wording in a few places. here.

avivace commented on src/HuC3.md:
```diff
@@ -0,0 +1,166 @@
+# HuC-3
+
+HuC-3 is an MBC used by some Game Boy games which, besides doing the
```
```suggestion
HuC-3 is an MBC used by some Game Boy games. Besides doing the
```

avivace commented on src/HuC3.md:
```diff
@@ -0,0 +1,166 @@
+# HuC-3
+
+HuC-3 is an MBC used by some Game Boy games which, besides doing the
+usual MBC stuff, also provides a real-time clock, speaker, and infrared
```
```suggestion
usual MBC stuff, it also provides a real-time clock, speaker, and infrared
```

cuavas replied:
I edited this paragraph to read better.

avivace commented on src/HuC3.md:
```diff
@@ -0,0 +1,166 @@
+  $D  | RTC semaphore (read/write)
+  $E  | IR (read/write)
+
+If any other value is written, reads from $A000–BFFF return open bus
```
Do we use "open bus values" anywhere else in the document? Or do we refer to them as rubbish etc? @ISSOtm

cuavas replied:
Why be less explicit?  “Rubbish” could be interpreted as meaning the chip drives the bus with an unpredictable value.  “Open bus” describes what actually happens with the hardware.

ISSOtm replied:
I'm not sure we use the term anywhere else, but yes, it's the correct term to use.

avivace commented on src/HuC3.md:
```diff
@@ -0,0 +1,166 @@
+
+### $E — IR [read/write]
+
+Similar to the equivalent register of the HuC1. The least significant
```
Maybe we can link to the "equivalent register of HuC1" here?

cuavas replied:
There’s already a link the the HuC1 page at the top.  It isn’t described very well there, anyway.

avivace commented on src/SUMMARY.md:
```diff
@@ -68,6 +68,7 @@
   - [MMM01](./MMM01.md)
   - [M161](./M161.md)
   - [HuC1](./HuC1.md)
+  - [HuC-3](./HuC3.md)
```
Maybe we could have a similar styling as HuC 1?
```suggestion
  - [HuC3](./HuC3.md)
```

cuavas replied:
The silkscreen on the HuC1 says “HuC1A” while the silkscreen on the HuC-3 says “HuC-3”.  The difference in punctuation is intentional.

avivace reviewed (changes requested):
Wow, thanks a lot @cuavas ! This looks already mostly merge-able, just left some comments

avivace commented on src/HuC1.md:
```diff
@@ -1,51 +1,53 @@
-I took a look, and it turns out that HuC1 differs from MBC1 quite a lot.
+HuC1 is an MBC developed by Hudson Soft. It implements ROM and RAM
+banking, and also provides infrared communication. Many sources on the
+internet said that HuC1 was “similar to MBC1”, without providing any
```
I would replace this with (starting with "Many sources on the", from the previous line)

```
Unlike generally believed, the HuC1 differs from MBC1 significantly.
```

cuavas replied:
But that reads poorly in English – if you really want to rework it in that direction, it should be “Contrary to popular belief…”  But then you’re not explaining where the popular belief came from (i.e. these repeated claims on various web sites).

Anyway, if you want to endlessly argue about wording like this, count me out.  I have more important things to do.  I just wanted to get write up my findings on the HuC-3 while it was fresh in my mind after doing reverse-engineering and emulation work on it.

ISSOtm replied:
I don't think the "many sources on the Internet" part belongs in Pan Docs at all, which would solve the problem.

> Anyway, if you want to endlessly argue about wording like this, count me out. I have more important things to do. I just wanted to get write up my findings on the HuC-3 while it was fresh in my mind after doing reverse-engineering and emulation work on it.

Considering that all of the technical tidbits have been approved by both of us, I'd entirely understand if you decline to participate in the grammar/presentation bikeshedding; we can make phrasing changes ourselves. Regardless, thank you *very much* for your research, and for caring enough to reshare it. Know that it's appreciated.

cuavas replied:
> I don't think the "many sources on the Internet" part belongs in Pan Docs at all, which would solve the problem.

Well, it gives historical context for JRRA’s comments, and it’s a sad fact that there are still many outdated websites that misleadingly say the HuC1 and HuC-3 are “similar to MBC1” or “similar to MBC5”.  It’s another thing carried over from the original blog post.

ISSOtm commented on src/HuC1.md:
```diff
@@ -1,51 +1,53 @@
-that HuC1 was \"similar to MBC1\", but they didn\'t provide any detail.
-I took a look, and it turns out that HuC1 differs from MBC1 quite a lot.
+HuC1 is an MBC developed by Hudson Soft. It implements ROM and RAM
+banking, and also provides infrared communication. Many sources on the
```
I disagree, the two consecutive "and"s would be confusing.

ISSOtm commented on src/HuC1.md:
```diff
@@ -1,51 +1,53 @@
-that HuC1 was \"similar to MBC1\", but they didn\'t provide any detail.
-I took a look, and it turns out that HuC1 differs from MBC1 quite a lot.
+HuC1 is an MBC developed by Hudson Soft. It implements ROM and RAM
+banking, and also provides infrared communication. Many sources on the
+internet said that HuC1 was “similar to MBC1”, without providing any
+detail. However, the HuC1 differs from MBC1 significantly.
```
```suggestion
banking, and also provides infrared communication.
Despite many sources on the internet claiming that HuC1 is “similar to MBC1”, it actually differs from MBC1 significantly.
```
Is this a reasonable compromise?

cuavas replied:
Yeah, that reads OK.
